In [1]:
import pysam
import numpy as np
import cPickle as pickle

In [2]:
%matplotlib inline
%load_ext autoreload
%autoreload 2
import matplotlib.pyplot as plt
from pyUMI import umi_stats as stats
from pyUMI import multimap_solver as solver

In [3]:
import logging
logging.basicConfig(level=logging.INFO)

In [4]:
from pyUMI import repeat_solver

In [5]:
#AAAACT  AAAGTT  ATATAG  ATCAAA  CCCTGG  CCGGAC  GACCGC  GGCGTC  GTTCGA  TAAAGT  TCTGCA  TTAATC

In [6]:
st = pysam.AlignmentFile("/data/UMI/data/MUS/sample_AAAACT.bam","rb")
#fst = repeat_solver.filter_alignmentFile(st, cell_barcode="TTAATC",repeats=True)

-----------------

In [37]:
st = pysam.AlignmentFile("/data/UMI/data/MUS/sample_AAAACT.bam","rb")

In [7]:
q = [k for k in stats.multimap_generator(st)]

In [20]:
g = [k.get_tag("GE") for k in q if k.has_tag("GE") and not "ENSMUS" in k.get_tag("GE") ]

In [21]:
from collections import Counter
gg = Counter(g)
len(gg)
#gg.most_common(10)

128363

In [22]:
gg

Counter({'B2_Mm2_dup42359': 8,
         'B2_Mm2_dup36715': 1,
         'L1Md_T_dup12440': 2,
         'B2_Mm2_dup36713': 9,
         'B1_Mus2_dup57778': 9,
         'B2_Mm2_dup36711': 2,
         'B2_Mm2_dup36710': 1,
         'L1Md_T_dup21944': 1,
         'B1_Mus2_dup57776': 1,
         'L1Md_T_dup21940': 3,
         'B2_Mm2_dup36719': 29,
         'RLTR31D_MM_dup384': 2,
         'L1Md_F2_dup5498': 3,
         'B1_Mus1_dup5959': 2,
         'L1Md_T_dup15798': 2,
         'L1Md_T_dup15797': 13,
         'L1Md_T_dup15793': 338,
         'L1Md_T_dup21949': 3,
         'B2_Mm2_dup2111': 12,
         'B2_Mm2_dup2112': 4,
         'B2_Mm2_dup31846': 9,
         'B1_Mus2_dup13587': 10,
         'B2_Mm2_dup31844': 1,
         'B2_Mm2_dup2116': 2,
         'B2_Mm2_dup2119': 19,
         'B2_Mm1t_dup14081': 6,
         'B2_Mm2_dup31849': 20,
         'L1_Mus3_dup12378': 9,
         'B2_Mm1a_dup6949': 3,
         'B2_Mm1a_dup6948': 5,
         'RSINE1_dup3457': 2,
         'B2_Mm1a_dup6945': 3

In [125]:
%%time
st.reset()

mapped = st.mapped
unmapped = st.unmapped
multimaps, uniques = stats.count_um(st)

CPU times: user 28.5 s, sys: 191 ms, total: 28.7 s
Wall time: 28.7 s


In [126]:
print mapped
print unmapped
print multimaps
print uniques

17752333
664548
14353994
3398339


18416881

In [48]:
mm_hashtable = solver.build_multimapping_hashtable(st)

In [38]:
uniqs_hashtable, uniqs_map = solver.build_uniques_associations(st, cell_barcodes=["AAAACT"])

In [110]:
umi = solver.query_uniques_hashtable(uniqs_hashtable, keyword='umi')

In [111]:
umi

[('AAAACT',
  {'TTATTTATTG': ['HISEQ:280:C9J9KANXX:2:2105:20334:47158',
    'HISEQ:280:C9J9KANXX:1:1114:16864:22165',
    'HISEQ:280:C9J9KANXX:1:2106:1284:98650'],
   'TTATTTATTC': ['HISEQ:280:C9J9KANXX:1:1308:9365:7277',
    'HISEQ:280:C9J9KANXX:1:1311:8577:98570',
    'HISEQ:280:C9J9KANXX:1:1315:18632:70730',
    'HISEQ:280:C9J9KANXX:1:2107:14582:32091',
    'HISEQ:280:C9J9KANXX:2:1210:10994:21248',
    'HISEQ:280:C9J9KANXX:2:1307:7084:46546',
    'HISEQ:280:C9J9KANXX:2:2209:18295:13032'],
   'GATGACATGA': ['HISEQ:280:C9J9KANXX:1:1214:15743:56097',
    'HISEQ:280:C9J9KANXX:1:1309:5076:3945',
    'HISEQ:280:C9J9KANXX:1:2107:7608:17266',
    'HISEQ:280:C9J9KANXX:2:2307:13992:36606',
    'HISEQ:280:C9J9KANXX:1:1103:15512:38142',
    'HISEQ:280:C9J9KANXX:1:1106:7755:20967',
    'HISEQ:280:C9J9KANXX:1:1108:6561:97557',
    'HISEQ:280:C9J9KANXX:1:1314:19252:23623',
    'HISEQ:280:C9J9KANXX:1:2102:18059:84812',
    'HISEQ:280:C9J9KANXX:1:2107:16599:97085',
    'HISEQ:280:C9J9KANXX:1:2207:14

In [94]:
mm_genes = solver.query_multimap_hashtable(mm_hashtable, keyword='gene')

In [103]:
qn = solver.query_multimap_hashtable(mm_hashtable, keyword='umi')

In [104]:
qn

['AGTGTCGAAA',
 'GGGGCTGGTG',
 'TCGGGGCTAG',
 'CGGGGTGGGC',
 'AGTTGTCCTA',
 'GTATGGAGCC',
 'TTTCGTTTTT',
 'CTAAATTAGC',
 'TGCAGGCTTG',
 'GGTCAGTGCG',
 'TTGTACTTTA',
 'AGAAGACGAC',
 'ACTAGATGTA',
 'TTATTGATTT',
 'GGAATAATAG',
 'TGCGTACAAT',
 'TAGAGCCCCG',
 'GTTTCCCTCT',
 'CGCCGTTTCG',
 'GTTTTCATGG',
 'GAAGTAGTAT',
 'CACGGGAGGT',
 'CTTGAGGAAT',
 'TCTTGGATTT',
 'CCTTTGGGAT',
 'AGATTGTCCG',
 'AACAATTTGC',
 'TCCAATGTTG',
 'GATATTTGGC',
 'AATTGCTGTT',
 'TTCACTAGGA',
 'CGGTAGTCTC',
 'AACTTACCTT',
 'GAGTTTTTAT',
 'GCTGGTAAAA',
 'AATATTTCTC',
 'GAACTCTGGA',
 'AAAGCTTTTT',
 'CTGTAGCTTA',
 'TGCGCCCAGT',
 'TAAACAAGGG',
 'GGTTGCGCTC',
 'CCTATGTAAT',
 'GGTGAGGTTC',
 'GGCCGAACGA',
 'CTGTTGTACT',
 'TGGTTATATA',
 'CAGTAAGAGT',
 'CGTATTTGCA',
 'CAAGAGACTT',
 'TGGTGCTCGT',
 'TGGTATTTCT',
 'GCATGTGCGT',
 'CTGGTGTAAA',
 'TGGGTGGGCT',
 'GGCAAGGGAA',
 'TTATTATTTA',
 'AACCACATGT',
 'TAATTGGTAG',
 'TTTCTCGCCG',
 'GGGGTGCAAC',
 'CGTTGCGCGT',
 'TCCATCTGCA',
 'GGGATATTCC',
 'CCCGACGGTA',
 'TAATGTTGTG',
 'CGTGTTAA

In [39]:
uniqs_genes = solver.get_uniquesmap_genes(uniqs_map)

In [88]:
len(uniqs_genes)

2726958

In [86]:
uniqs_regions = solver.get_uniquesmap_genes(uniqs_map, gene=False)

In [89]:
len(uniqs_regions)

3398339

In [42]:
uniqs_repeats = [a for a in uniqs_genes if not("ENS" in a)]

In [92]:
len(solver.get_bamfile_genes(st))

12589260

In [46]:
len(genes)

12589260

In [44]:
len(uniqs_repeats)

194052

In [92]:
m1 = solver.build_multimapping_hashtable(st1)
uh1,u1 = solver.build_uniques_associations(st1, cell_barcodes=["AAAACT"])

In [16]:
len(u1)

3398339

In [7]:
fst1 = repeat_solver.filter_alignmentFile(st1, cell_barcode="U",repeats=True)
fst1 = pysam.AlignmentFile("/data/UMI/data/MUS/sample_U_filtered.bam","rb")

In [8]:
type(fst1)

pysam.calignmentfile.AlignmentFile

In [12]:
!pwd

/home/parastou/UMI


In [12]:
fst1.reset()
reads =  fst1.fetch(until_eof=True)

In [13]:
tags =[]
for read in reads:
    tags.append(read.get_tag("GE"))

In [14]:
tags[:100]

['B1_Mus2_dup41322',
 'B1_Mus1_dup56899',
 'B1_Mus1_dup56901',
 'B1_Mus1_dup56901',
 'B1_Mus1_dup56901',
 'L1VL1_dup1265',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup539',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup541',
 'MMETn-int_dup5

In [15]:
uh1_r,u1_r = solver.build_uniques_associations(fst1, cell_barcodes=["AAAACT"])

In [24]:
u1_r

{'HISEQ:280:C9J9KANXX:2:2316:6023:81395': ('B1_Mus1_dup38003',
  'chr7',
  123158638),
 'HISEQ:280:C9J9KANXX:2:1113:17108:20780': ('B2_Mm2_dup22563',
  'chr5',
  23714167),
 'HISEQ:280:C9J9KANXX:2:1314:20023:3601': ('B3_dup33937', 'chr4', 127240787),
 'HISEQ:280:C9J9KANXX:2:1303:8293:93992': ('B1F2_dup10270', 'chr10', 81257027),
 'HISEQ:280:C9J9KANXX:2:2204:11792:97321': ('B2_Mm2_dup75909',
  'chr15',
  65618767),
 'HISEQ:280:C9J9KANXX:2:2304:6092:44985': ('B2_Mm1t_dup7644',
  'chr6',
  39365825),
 'HISEQ:280:C9J9KANXX:1:1114:20764:69622': ('L1_Mus3_dup11813',
  'chr4',
  146516442),
 'HISEQ:280:C9J9KANXX:1:2212:14273:54606': ('Lx_dup12008', 'chrX', 15306901),
 'HISEQ:280:C9J9KANXX:1:1215:20953:38024': ('L1_Mur3_dup7282',
  'chr4',
  111843178),
 'HISEQ:280:C9J9KANXX:2:2211:13511:31107': ('B2_Mm1a_dup11770',
  'chr11',
  67137480),
 'HISEQ:280:C9J9KANXX:1:1302:7918:48028': ('ERVB5_1-I_MM-int_dup226',
  'chr13',
  100415961),
 'HISEQ:280:C9J9KANXX:1:2203:6312:3946': ('B2_Mm1t_dup3828', 

In [25]:
ur_tags = [u1_r[k][0] for k in u1_r]


In [ ]:
solver.get_uniquesmap_genes(uniques)

In [27]:
len(ur_tags)

194052

In [37]:
from collections import Counter
urt = Counter(ur_tags)
len(urt)
urt_list = [k for k in urt]

In [38]:
len(urt_list)

31416

In [40]:
ranking = []
st1.reset()
reads = st1.fetch(until_eof=True)

for read in reads:
    if (read.has_tag("GE")):
        ranking.append(read.get_tag("GE"))

In [56]:
ranking_dict = Counter(ranking)

In [59]:
import operator

ranking_dict_sorted = sorted(ranking_dict.items(), key=operator.itemgetter(1),reverse=True)


In [61]:
#ranking_dict_sorted
ranked_repeats = [(i,j) for (i,j) in ranking_dict_sorted if not("ENSMUS" in i)]

In [66]:
ranked_repeats

[('L1_Mus2_dup2959', 15829),
 ('L1_Mus1_dup17730', 11470),
 ('B2_Mm1a_dup7097', 8880),
 ('B2_Mm1t_dup14771', 7747),
 ('B2_Mm1a_dup7601', 7650),
 ('B2_Mm1a_dup17081', 7644),
 ('B2_Mm1a_dup1597', 7276),
 ('B2_Mm1a_dup14149', 7235),
 ('B2_Mm1a_dup1166', 7153),
 ('B2_Mm1a_dup16618', 7122),
 ('B2_Mm1a_dup16891', 7080),
 ('B2_Mm1t_dup1422', 7035),
 ('B2_Mm1t_dup13627', 7033),
 ('B2_Mm1t_dup1981', 7007),
 ('B2_Mm1a_dup16046', 6993),
 ('B2_Mm1a_dup7071', 6955),
 ('B2_Mm1t_dup17631', 6915),
 ('B2_Mm1a_dup15240', 6762),
 ('L1Md_T_dup21899', 6690),
 ('B2_Mm1a_dup14394', 6393),
 ('B2_Mm1a_dup11250', 6197),
 ('B2_Mm1a_dup763', 6091),
 ('B2_Mm1a_dup13412', 6088),
 ('B2_Mm1a_dup6277', 6004),
 ('L1_Mus3_dup13458', 6003),
 ('B2_Mm2_dup75868', 5945),
 ('B2_Mm1t_dup2319', 5895),
 ('RLTR19-int_dup2148', 5859),
 ('B2_Mm1t_dup18156', 5782),
 ('B2_Mm1a_dup7492', 5689),
 ('B2_Mm1a_dup9587', 5677),
 ('B2_Mm1a_dup5014', 5613),
 ('Lx5_dup3006', 5496),
 ('B2_Mm1a_dup12483', 5411),
 ('B2_Mm1a_dup3351', 5362),
 ('R

In [69]:
ranked_unique_repeats = sorted(urt.items(), key=operator.itemgetter(1),reverse=True)

In [82]:
ranked_unique_repeats

[('B2_Mm1a_dup12765', 1484),
 ('B2_Mm1a_dup17081', 1309),
 ('B2_Mm1t_dup4289', 1124),
 ('L1_Mur2_dup15158', 1015),
 ('PB1D7_dup8929', 939),
 ('RMER15_dup16209', 913),
 ('B2_Mm1a_dup16722', 902),
 ('B1_Mus2_dup22174', 750),
 ('L1_Mus1_dup6883', 682),
 ('B2_Mm1t_dup16939', 678),
 ('B2_Mm1a_dup8549', 567),
 ('B2_Mm1t_dup7380', 551),
 ('L1_Mus3_dup4038', 485),
 ('L1Md_F2_dup63890', 445),
 ('RLTR25B_dup2452', 432),
 ('B2_Mm2_dup78396', 413),
 ('B1_Mur2_dup6022', 402),
 ('ORR1C2_dup2292', 380),
 ('B3_dup105592', 347),
 ('L1Md_T_dup16347', 327),
 ('B2_Mm2_dup62345', 320),
 ('B2_Mm1t_dup10146', 314),
 ('L1VL4_dup5584', 313),
 ('Lx3B_dup2711', 297),
 ('B1_Mus1_dup29879', 277),
 ('B2_Mm1t_dup18687', 252),
 ('L1_Mm_dup4278', 248),
 ('B2_Mm2_dup3439', 240),
 ('L1Md_Gf_dup537', 237),
 ('L1MEd_dup2089', 233),
 ('B2_Mm1a_dup5210', 230),
 ('L1Md_T_dup8189', 226),
 ('ORR1B2_dup3730', 224),
 ('L1_Mur3_dup5547', 223),
 ('B2_Mm2_dup86087', 222),
 ('Lx2_dup1313', 220),
 ('L1_Mus1_dup23474', 208),
 ('B3_dup

    Gene                      Uniquely mapped reads
    'B2_Mm1a_dup12765'        1484
    'B2_Mm1a_dup17081'        1309
    'B2_Mm1t_dup4289'         1124
    'L1_Mur2_dup15158'        1015
    'PB1D7_dup8929'           939

In [86]:
with open('/data/UMI/data/MUS/AAAACT_ranked_unique_repeats.pkl', 'wb') as handle:
    pickle.dump(ranked_unique_repeats, handle, protocol=pickle.HIGHEST_PROTOCOL)

In [72]:
unique_repeats = [k for k in urt]

In [75]:
top_counter = 5
for item in ranked_repeats:
    if item[0] in unique_repeats:
        print item
        top_counter -= 1
    if not(top_counter):
        break

('L1_Mus2_dup2959', 15829)
('L1_Mus1_dup17730', 11470)
('B2_Mm1a_dup7097', 8880)
('B2_Mm1t_dup14771', 7747)
('B2_Mm1a_dup7601', 7650)


    Gene                  Frequency    Uniques

    'L1_Mus2_dup2959',    15829        5
    'L1_Mus1_dup17730'    11470        76
    'B2_Mm1a_dup7097'     8880         1
    'B2_Mm1t_dup14771'    7747         1
    'B2_Mm1a_dup7601'     7650         3

-----------------------------------------

    ERVB4_2-LTR_MM_dup63

    ERVB4_2-LTR_MM_dup63    chrY    +    29718275    29718800

    Read length = 50bp

    Read name = HISEQ:280:C9J9KANXX:2:2115:15792:76261

    ('ERVB4_2-LTR_MM_dup63', 'chrY', 29718465)

    Location = chrY:29,718,472

    Read name = HISEQ:280:C9J9KANXX:1:1310:7181:38503

    ('ERVB4_2-LTR_MM_dup63', 'chrY', 29718496)

    Location = chrY:29,718,546

    275 800

    472 522

    546 596

